### PR TITLE
Fixes an issue where the queue was still being fetched if the app was running in background.

### PR DESCRIPTION
**Pull request by BernardGatt**
Full details can be found here: https://github.com/customerio/issues/issues/9048

I'm not sure if this would create any side effects with Scenes. Can anyone confirm?

### CODE REVIEW

File: Gist/Managers/MessageQueueManager.swift [hunk 1]
```diff
@@ -2,61 +2,69 @@ import Foundation
...
-                        let position = message.gistProperties.position
-
-                        if let routeRule = message.gistProperties.routeRule {
-                            let cleanRouteRule = routeRule.replacingOccurrences(of: "\\", with: "/")
-                            if let regex = try? NSRegularExpression(pattern: cleanRouteRule) {
-                                let range = NSRange(location: 0, length: Gist.shared.getCurrentRoute().utf16.count)
-                                if regex.firstMatch(in: Gist.shared.getCurrentRoute(), options: [], range: range) == nil {
-                                    Logger.instance.debug(message:
-                                        "Current route is \(Gist.shared.getCurrentRoute()), needed \(cleanRouteRule)")
+        if UIApplication.shared.applicationState != .background {
```
Comment (levibostian) [inline Gist/Managers/MessageQueueManager.swift:22]: I would test this code under an iOS app that uses Scenes and a separate test for a separate app using SwiftUI. 

I am not very experienced with Scene apps, but with my SwiftUI experience, if you have built an app using SwiftUI and no longer have an `AppDelegate` file, many functions that you call against `UIApplication.shared` do not work as expected. 

So, I am thinking that *maybe* a SwiftUI app will make this line of code always return false.

File: Gist/Managers/MessageQueueManager.swift [hunk 2]
```diff
@@ -2,61 +2,69 @@ import Foundation
...
+    
     func setup() {
         queueTimer = Timer.scheduledTimer(timeInterval: 10,
                                           target: self,
                                           selector: #selector(checkForMessages),
                                           userInfo: nil,
                                           repeats: true)
-        checkForMessages()
+        
+        // Wait 1 second for app to become active.
```
Comment (levibostian) [inline Gist/Managers/MessageQueueManager.swift]: I want to suggest adding a little more notes to the comment suggesting why this change was introduced. 

What was happening before that adding a delay helps resolve?

File: Gist/Managers/MessageQueueManager.swift (hunk 1 above)
Comment (BernardGatt) [reply Gist/Managers/MessageQueueManager.swift:22]: That's my main worry, we will need to test this fix on a pure SwiftUI app

File: Gist/Managers/MessageQueueManager.swift (hunk 2 above)
Comment (BernardGatt) [reply Gist/Managers/MessageQueueManager.swift]: I added more context in the comments.

On application launch, there's a short period where the app still shows as in "background" and since we're now checking the state before fetching the queue, the first check was always skipped.

File: Gist/Managers/MessageQueueManager.swift (hunk 1 above)
Comment (levibostian) [reply Gist/Managers/MessageQueueManager.swift:22]: I did a quick check to see if I could find anything on a SwiftUI method to detect application state. My thought is that you could change your if logic to check for a UIKit value and a SwiftUI kit value to see if either of them suggest the app is in the foreground and if so, run the queue. 

Many suggestions I found online were about how to *observe* the app state (not a surprise since it's SwiftUI). I *think* I found something that isn't observing, but just a synchronous check for the application state. 

I used [this article](https://betterprogramming.pub/swiftui-tips-detecting-a-swiftui-apps-active-inactive-and-background-state-a5ff8acf5db1) for inspiration which led me to:

```swift
Environment(EnvironmentValues.scenePhase).wrappedValue
```

Taking [the init function of `Environment`](https://developer.apple.com/documentation/swiftui/environment/init(_:)) to read [the app state](https://developer.apple.com/documentation/swiftui/scenephase) via [`EnvironmentValues`](https://developer.apple.com/documentation/swiftui/environmentvalues).

Comment (BernardGatt) [reply Gist/Managers/MessageQueueManager.swift:22]: Looking at Apple's [documentation](https://developer.apple.com/documentation/uikit/uiapplication/1623003-applicationstate) about this subject, it looks like what we have should be fine already:

> In a scene-based app, this property takes the value of the most active scene, which it determines from each scene’s [activationState](https://developer.apple.com/documentation/uikit/uiscene/3197908-activationstate) property.

The demo app uses scenes (thinking about it now, that's why I had to add the delay), I'm going to test on a pure "no app delegate" iOS app and revert back, that's the only way to confirm.

Comment (BernardGatt) [reply Gist/Managers/MessageQueueManager.swift:22]: Hey @levibostian, I created a new, from scratch Swift UI demo app, initialized the SDK in the new "App" file.

```swift
import SwiftUI
import Gist

@main
struct PureSwiftUITestApp: App {
    init() {
        Gist.shared.setup(siteId: "a5ec106751ef4b34a0b9", dataCenter: "eu", logging: true, env: .production)
        Gist.shared.setUserToken("ABC123")
    }
    
    var body: some Scene {
        WindowGroup {
            ContentView()
        }
    }
}
```

I confirm that the fix is working as expected, the queue stops when the app is in the background and resumes when the app is active again.

Shall we merge this?

File: Gist/Managers/MessageQueueManager.swift [hunk 3]
```diff
@@ -2,61 +2,70 @@ import Foundation
...
                                           selector: #selector(checkForMessages),
                                           userInfo: nil,
                                           repeats: true)
-        checkForMessages()
+        
+        // Since on app launch there's a short period where the applicationState is still set to "background"
+        // We wait 1 second for the app to become active before checking for messages.
+        DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
+            self.checkForMessages()
+        }
```
Comment (Shahroz16) [inline Gist/Managers/MessageQueueManager.swift:17]: Out of curiosity, is there a way, we can get an instance of the `scene phase` that we can observe instead of a 1-second check? 

I know in swift-ui we can also observe notification center to know if the app is in foreground or background, maybe we can use that as a trigger to start/stop synching?

Comment (BernardGatt) [reply Gist/Managers/MessageQueueManager.swift:17]: It should be possible to observe changes, at this point we're only using this to make sure the app has transitioned into active in case Gist is initialized the first thing on start.

If we eventually go down this route we can use the notification to actually start and stop the queue listener as well.

**Review by Shahroz16 — APPROVED** [review]

**Review by levibostian — APPROVED** [review]
